2016-2017_Starstruck/Module Test Code/FileIO_Test/src/AutonRecorder.c

```diff
@@ -0,0 +1,193 @@
+       joystickGetDigital(1,7,JOY_UP),joystickGetDigital(1,7,JOY_DOWN),joystickGetDigital(1,7,JOY_LEFT),joystickGetDigital(1,7,JOY_RIGHT),
+       joystickGetDigital(1,8,JOY_UP),joystickGetDigital(1,8,JOY_DOWN),joystickGetDigital(1,8,JOY_LEFT),joystickGetDigital(1,8,JOY_RIGHT)},
+      {joystickGetAnalog(2,1),joystickGetAnalog(2,2),joystickGetAnalog(2,3),joystickGetAnalog(2,4)},
+      {joystickGetDigital(2,5,JOY_UP),joystickGetDigital(2,5,JOY_DOWN),joystickGetDigital(2,6,JOY_UP),joystickGetDigital(2,6,JOY_DOWN),
+       joystickGetDigital(2,7,JOY_UP),joystickGetDigital(2,7,JOY_DOWN),joystickGetDigital(2,7,JOY_LEFT),joystickGetDigital(2,7,JOY_RIGHT),
+       joystickGetDigital(2,8,JOY_UP),joystickGetDigital(2,8,JOY_DOWN),joystickGetDigital(2,8,JOY_LEFT),joystickGetDigital(2,8,JOY_RIGHT)},
+       NULL,NULL
+    };
+    frame.next=&frame;
+    frame.previous=&frame;
+    return frame;
+  }
+
+  void addFrame(Frame toAdd){
+      toAdd.previous = root.previous;
+      toAdd.next = &root;
+      root.previous = &toAdd;
+  }
+
+  char * frameToString(Frame frame){
+    char * string = malloc(50);
+    sprintf(string,"%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d%d",
+      frame.analog_main[0]+127,
+      frame.analog_main[1]+127,
+      frame.analog_main[2]+127,
+      frame.analog_main[3]+127,
+      frame.digital_main[0],
+      frame.digital_main[1],
+      frame.digital_main[2],
+      frame.digital_main[3],
+      frame.digital_main[4],
+      frame.digital_main[5],
+      frame.digital_main[6],
+      frame.digital_main[7],
+      frame.digital_main[8],
+      frame.digital_main[9],
+      frame.digital_main[10],
+      frame.digital_main[11],
+      frame.analog_partner[0]+127,
+      frame.analog_partner[1]+127,
+      frame.analog_partner[2]+127,
+      frame.analog_partner[3]+127,
+      frame.digital_partner[0],
+      frame.digital_partner[1],
+      frame.digital_partner[2],
+      frame.digital_partner[3],
+      frame.digital_partner[4],
+      frame.digital_partner[5],
+      frame.digital_partner[6],
+      frame.digital_partner[7],
+      frame.digital_partner[8],
+      frame.digital_partner[9],
+      frame.digital_partner[10],
+      frame.digital_partner[11]
+    );
+    // TODO: code above is really ugly and inefficient lol
+    printf("%s\n",string);
+    return string;
+  }
+
+  Frame stringToFrame(char* string){
+    Frame frame = NULLFRAME;
+    char* analog_main = substring(string,1,12);
+    char* digital_main = substring(string,13,12);
+    char* analog_partner = substring(string,25,12);
+    char* digital_partner = substring(string,37,12);
+
+    // printf("%d\t%d\t%d\t%d\t",frame.analog_main[0],frame.analog_main[1],frame.analog_main[2],frame.analog_main[3]);
+
+    int x;
+    for(x=0;x<4;x++){
+      frame.analog_main[x] = atoi(substring(analog_main,(x*3)+1,3))-127;
```

Comment:
might want to add comments to explain this math cuz we're not gonna remember what it means later